scripts/stubsabot.py

```diff
@@ -170,21 +177,72 @@ def __str__(self) -> str:
         return f"Skipping {self.distribution}: {self.reason}"
 
 
+def wheel_dir_is_py_typed(directory: zipfile.Path, *, dist_name: str) -> bool:
+    top_level = list(directory.iterdir())
+    if any(path.is_dir() and path.name.endswith(".dist-info") for path in top_level):
+        raise RuntimeError(f"Unexpectedly found a `.dist-info` subdirectory in the wheel for {dist_name}!")
+    if all(path.is_dir() for path in top_level):
+        # The directory only contains other directories: we're probably looking at a namespace package
+        # Recurse into the lower-down directories to see if they all have py.typed files.
+        return all(wheel_dir_is_py_typed(path, dist_name=dist_name) for path in top_level)
+    return any(path.is_file() and path.name == "py.typed" for path in top_level)
+
+
+def all_pkgs_in_wheel_are_py_typed(wheel_file: zipfile.ZipFile, *, dist_name: str) -> bool:
+    wheel = zipfile.Path(wheel_file)
+    top_level_contents = [path for path in wheel.iterdir() if not (path.is_dir() and path.name.endswith(".dist-info"))]
+    if not top_level_contents:
+        warnings.warn(f"Unexpectedly found an empty wheel for {dist_name}", EmptyRuntimeDistributionWarning, stacklevel=1)
```

Comment:
originally I had this as an exception, but it raised for TensorFlow! That was pretty surprising to me (and might indicate a bug somewhere else in stubsabot? Maybe we shouldn't be looking at this wheel if it's empty?), so I opted to emit a warning for now.